### DISCUSSION TITLE
Why can't I get injections working?

### DISCUSSION THREAD

**Discussion by Hubro** [nvim-treesitter/nvim-treesitter#5690] [Q&A] [answered]
I'm using https://github.com/IndianBoy42/tree-sitter-just, and the injection queries don't work by default. I thought I'd fix it real quick, and 2 hours later I'm tearing my hair out.

I have this simple query in `queries/just/injections.scm`:

```scm
(shebang_recipe
  (shebang_body) @injection.content (#set! injection.language "bash"))
```

I have also tried setting it directly with Lua:

```lua
vim.treesitter.query.set("just", "injections", [[
(shebang_recipe
  (shebang_body) @injection.content (#set! injection.language "bash"))
]])
```

I have this highlight query in `queries/just/highlights.scm`, as instructed by CONTRIBUTING.md, and it works:

```scm
(shebang_recipe
  (shebang interpreter: (TEXT) @keyword) @comment
  (shebang_body) @text.literal.block (#set! "priority" 90))
```

I have tested the injections query in the `:EditQuery` window and confirmed it captures the right text:

![image](https://github.com/neovim/neovim/assets/597206/501c1ce2-cbdc-4307-98fc-2cba01609746)

I've tried explicitly setting the injection language to "sh" and "bash", but it doesn't freaking work. Nothing happens, I may as well not have the injection query defined. As you see in the screenshot everything is just green, clearly not highlighted as bash.

I've gone through a bunch of other injection queries in this repo to see if I missed anything.

I've tested latest nvim v0.9.4 and latest nightly, and I'm using the most up-to-date master branch of nvim-treesitter.

I have followed all the available instructions and I have no clue where to start looking to figure out why this isn't working. ChatGPT is completely stumped as well.

Any advice would be appreciated.

---

EDIT: If relevant, "just" hasn't been added to nvim-treesitter, so I've manually added it:

```lua
local parsers = require("nvim-treesitter.parsers")
local configs = parsers.get_parser_configs()

configs.just = {
  install_info = {
    url = "https://github.com/IndianBoy42/tree-sitter-just",
    files = { "src/parser.c", "src/scanner.cc" },
    branch = "main",
  },
  filetype = "just",
}
```

**Comment by clason** [comment]
> ChatGPT is completely stumped as well.

Well, yes; that will lie blindly to your face and gaslight you about it afterwards. Not sure why you expected factual information.

Did you in fact install the parsers for the languages you want to inject?

**Reply by Hubro** [reply]
> Did you in fact install the parsers for the languages you want to inject?

Yep, triple checked that as well, forgot to mention it.

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/597206/674c4c95-aad2-476d-9e34-3845a1d53d0a)

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/597206/72024bdb-39a3-4037-82f4-0c406c8f2149)

**Reply by Hubro** [reply]
I only mentioned ChatGPT jokingly, it's usually very little help with this kind of niche issue :smile:

**Comment by amaanq** [comment]
I tried it out myself, it really doesn't seem to work for some odd reason, even with this simple query that *should* work:
```scm
((shebang_body) @injection.content
  (#set! injection.combined)
  (#set! injection.language "bash"))
```

I debugged languagetree.lua a bit and can see the ranges are fetched accordingly for the injection in _get_injection, but I get a bit lost after that, maybe Lewis would know more

**Reply by Hubro** [reply]
Did you add the parser through your config in the same way I did?

I'm wondering if maybe there's a bug that prevents injections from working when the parser config isn't built-in. That's the only difference I see between Just and all the other parsers that work just fine.

**Reply by clason** [reply, marked as answer]
No, that should make zero difference (unless you are doing some sort of lazy-loading -- which you shouldn't).

I'd rather suspect something like `injection.include-children` to make a difference.

**Reply by Hubro** [reply]
@clason You're right, I forked nvim-treesitter and added the parser "properly". The only difference I found is that opening a justfile now crashes violently and syntax highlighting is broken, because I haven't installed the bash parser yet:

```
Error in decoration provider treesitter/highlighter.win:
Error executing lua: ...-linux64/share/nvim/runtime/lua/vim/treesitter/query.lua:248: Query error at 9:4. Invalid node type "heredoc_end":
  (heredoc_end) @injection.language)
   ^

stack traceback:
        [C]: in function '_ts_parse_query'
        ...-linux64/share/nvim/runtime/lua/vim/treesitter/query.lua:248: in function 'fn'
        ...vim-linux64/share/nvim/runtime/lua/vim/func/_memoize.lua:54: in function 'fn'
        ...vim-linux64/share/nvim/runtime/lua/vim/func/_memoize.lua:54: in function 'get'
        ...4/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:132: in function 'new'
        ...4/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:503: in function 'add_child'
        ...4/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:388: in function '_add_injections'
        ...4/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:440: in function 'parse'
        ...64/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:328: in function <...64/share/nvim/runtime/lua/vim/treesitter/highl
```

Installing the bash parser fixes it.

At least this confirms that the injections do *something*, otherwise the bash parser wouldn't have been necessary. Still no clue why I don't get syntax highlighting though.

**Reply by Hubro** [reply]
@clason Hey, `injection.include-children` worked...

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/597206/6495da23-46f1-40b7-a7bc-3f62d178008f)

I feel like I should be happy, but the fact that this wasn't mentioned in CONTRIBUTING.md has completely wasted over 4 hours of my life... All this troubleshooting and all along I was just missing 1 undocumented directive. :sob:

**Reply by clason** [reply]
> I feel like I should be happy, but the fact that this wasn't mentioned in CONTRIBUTING.md has completely wasted over 4 hours of my life... All this troubleshooting and all along I was just missing 1 undocumented directive. 😭


You are (still) missing the fact that highlighting and injections are completely done by Neovim, not by nvim-treesitter -- and hence the relevant documentation (which obviously has room for improvement) is in `:h treesitter`.

I remind that treesitter support (and nvim-treesitter) is still experimental and assumes quite a bit of familiarity with all the moving parts.

**Reply by amaanq** [reply]
Right I mixed up combined and include-children, oops :sweat_smile:
